src/util/customErrors.ts

```diff
@@ -0,0 +1,25 @@
+/**
+ * A custom error that holds additional information and a more explicit message
+ * to simplify tracking down errors that occur during execution
+ */
+export class AnnotatedError extends Error {
+  // TODO: need to explicitly specify cause in TypeScript versions < 4.7
+  // Remove this when TypeScript is upgraded, and add the cause directly to the `super` call
+  cause: Error;
+
+  constructor(
+    originalError: Error,
+    message: string,
+    expressionName: string,
+    libraryName: string,
+    localId?: string
```

Comment:
Suggest adding the `locator` as a possible argument to AnnotatedError. This would allow a consumer of the error to print out the CQL around the error and highlight that clause that had the issue.

Reply:
This is super nice. It requires adding a `locator` property on the root `Expression` class constructor, but I don't see any downside in doing that as long as the error reporting behaves properly when locators aren't there (similar to when localIds aren't there.

@cmoesel are you good with this potential change? basically:

```ts
// In Expression constructor
if (json.locator != null) {
  this.locator = json.locator;
}
```

and then rendering it in annotated errors:

```
        Error Message:  you messed up 
        CQL Library:    lib|5.0.000
        Expression:     CalculateAgeAt
        ELM Local ID:   28
        CQL Locator:    35:29-35:86
```

Reply:
@hossenlopp You mentioned having consumers be able to utilize the information. This made me wonder if the `AnnotatedError` should also contain this rendered info in a structured way to potentially consume elsewhere?

```ts
try {
  // something
} catch(e: unknown) {
  if (e instanceof AnnotatedError /* imported from cql-execution? */) {
    // e has the following properties
    e.expressionName;
    e.libraryName;
    e.localId;
    e.locator;
  }
}
```

Could be overkill. Just a thought

Reply:
Isn't this data included in the error that gets thrown? But yeah exporting the type would allow it to be used without crazy shenanigans.

Reply:
Technically it isn't, it's just placed directly in the error message. If we make the constructor args actually persist on the class instance (like using the `public` keyword in TS), then they will be available:

![image](https://user-images.githubusercontent.com/16297930/224837249-dec01c70-20b1-4e92-a0e8-9e17cfe712e2.png)

Reply:
I completely missed that they were not being persisted. But I do see benefit in doing so.

Reply:
I'm down w/ all of this (locators in the constructor, stored properties, etc.).

Teamwork is the dream work!